SIGSEGV dockerd 20.10.2

* [x] This is a bug report
* [ ] This is a feature request
* [x] I searched existing issues before opening this one

### Expected behavior
dockerd must never terminate with SIGSEGV.

### Actual behavior
`/usr/bin/dockerd -D`

```
INFO[2021-01-16T10:41:01.251777614+01:00] Starting up                                  
DEBU[2021-01-16T10:41:01.252491225+01:00] Listener created for HTTP on unix (/var/run/docker.sock) 
DEBU[2021-01-16T10:41:01.253124266+01:00] Golang's threads limit set to 229950         
INFO[2021-01-16T10:41:01.253558556+01:00] parsed scheme: "unix"                         module=grpc
INFO[2021-01-16T10:41:01.253581766+01:00] scheme "unix" not registered, fallback to default scheme  module=grpc
DEBU[2021-01-16T10:41:01.253609706+01:00] metrics API listening on /var/run/docker/metrics.sock 
INFO[2021-01-16T10:41:01.253624336+01:00] ccResolverWrapper: sending update to cc: {[{unix:///run/containerd/containerd.sock  <nil> 0 <nil>}] <nil> <nil>}  module=grpc
INFO[2021-01-16T10:41:01.253712596+01:00] ClientConn switching balancer to "pick_first"  module=grpc
INFO[2021-01-16T10:41:01.254819988+01:00] parsed scheme: "unix"                         module=grpc
INFO[2021-01-16T10:41:01.254857268+01:00] scheme "unix" not registered, fallback to default scheme  module=grpc
INFO[2021-01-16T10:41:01.254878428+01:00] ccResolverWrapper: sending update to cc: {[{unix:///run/containerd/containerd.sock  <nil> 0 <nil>}] <nil> <nil>}  module=grpc
INFO[2021-01-16T10:41:01.254891448+01:00] ClientConn switching balancer to "pick_first"  module=grpc
DEBU[2021-01-16T10:41:01.255547058+01:00] Using default logging driver json-file       
DEBU[2021-01-16T10:41:01.255606688+01:00] [graphdriver] priority list: [btrfs zfs overlay2 fuse-overlayfs aufs overlay devicemapper vfs] 
INFO[2021-01-16T10:41:01.255636908+01:00] [graphdriver] using prior storage driver: btrfs 
DEBU[2021-01-16T10:41:01.255663819+01:00] Initialized graph driver btrfs               
DEBU[2021-01-16T10:41:01.255673489+01:00] processing event stream                       module=libcontainerd namespace=plugins.moby
DEBU[2021-01-16T10:41:01.929682530+01:00] No quota support for local volumes in /var/lib/docker/volumes: Filesystem does not support, or has not enabled quotas 
WARN[2021-01-16T10:41:01.949370910+01:00] Your kernel does not support CPU realtime scheduler 
WARN[2021-01-16T10:41:01.949392680+01:00] Your kernel does not support cgroup blkio weight 
WARN[2021-01-16T10:41:01.949400020+01:00] Your kernel does not support cgroup blkio weight_device 
DEBU[2021-01-16T10:41:01.949506261+01:00] Max Concurrent Downloads: 3                  
DEBU[2021-01-16T10:41:01.949514651+01:00] Max Concurrent Uploads: 5                    
DEBU[2021-01-16T10:41:01.949519791+01:00] Max Download Attempts: 5                     
INFO[2021-01-16T10:41:01.949532081+01:00] Loading containers: start.                   
DEBU[2021-01-16T10:41:01.949707031+01:00] processing event stream                       module=libcontainerd namespace=moby
ERRO[2021-01-16T10:41:01.951088322+01:00] failed to load container                      container=9dd8da6d68e1dbe5407fdde3619f9f23319c8ad864d270602c822c2392f52927 error="open /var/lib/docker/containers/9dd8da6d68e1dbe5407fdde3619f9f23319c8ad864d270602c822c2392f52927/config.v2.json: no such file or directory"
DEBU[2021-01-16T10:41:01.955084726+01:00] loaded container                              container=cbe36a4aedddad663c47779dcb8abc2dc7843a38323d35c23df2886ef8564640 paused=false running=true
DEBU[2021-01-16T10:41:01.955085426+01:00] loaded container                              container=94a817c8f17e4a332c108c2afb9b10b4fc0577177ae73b82cb234b5917c7c41b paused=false running=true
DEBU[2021-01-16T10:41:01.955094976+01:00] loaded container                              container=b9bfe91a927adac0f9af1d0291374cf004b1045ec0c8d1186265aa579f2babab paused=false running=true
DEBU[2021-01-16T10:41:01.955094916+01:00] loaded container                              container=4bf8dc61c55d5b8dfee5d1390184f0eb21812595178a18f2657d3c8693eaafba paused=false running=true
DEBU[2021-01-16T10:41:01.955097796+01:00] loaded container                              container=dfc9cf08ad34f70818a6accf5f8db51d7568f9b255068933870a2a932215a01c paused=false running=true
DEBU[2021-01-16T10:41:01.955171516+01:00] loaded container                              container=fe9e1626adf23de5df53e86d3b057cc6dfda72ed3eb916bafe76492223475a8f paused=false running=true
DEBU[2021-01-16T10:41:01.955124196+01:00] loaded container                              container=3b3f5d0401938e1a3d99df9ef8e3fdb7027c0bb898cbddbf6f37b50dd17fc944 paused=false running=true
DEBU[2021-01-16T10:41:01.955124056+01:00] loaded container                              container=77254d04500e50148af90e105ed5d7f8bc7a4e5303330e6d28c915ec54055e26 paused=false running=true
DEBU[2021-01-16T10:41:01.955127846+01:00] loaded container                              container=75779d522db55c31da24af99d985917f34379311514d951798cd2694c6aa6025 paused=false running=true
DEBU[2021-01-16T10:41:01.955175656+01:00] loaded container                              container=58f48df9ab4e7e473ac00b3dc5c665fa8b0181dade8b7e78d62e49882a89b7fb paused=false running=true
DEBU[2021-01-16T10:41:01.955213656+01:00] loaded container                              container=6b5fd16689b484e04491b5d35f6fcb5f89c30f91a0e82b1c0514bd5c6376a484 paused=false running=true
DEBU[2021-01-16T10:41:01.955265086+01:00] loaded container                              container=6c58b36126694662cce12e48e071d74e5c5236f662a592ad272e6d1701df4245 paused=false running=true
DEBU[2021-01-16T10:41:01.955323526+01:00] loaded container                              container=1d587fc235238bff73224eef576b6acd63d1e0dc42ce07cb32c53916ddb7b1ec paused=false running=true
DEBU[2021-01-16T10:41:01.955357136+01:00] loaded container                              container=3003e2bfed86be6550eda431ecae96cca300de0de5db6bb56864b65147f79dd6 paused=false running=true
DEBU[2021-01-16T10:41:01.955416256+01:00] loaded container                              container=ea9c118cc1d62286daccf23a4db5b7cc3a9eae78527a920ac33febffbfd924a6 paused=false running=true
DEBU[2021-01-16T10:41:01.955452696+01:00] loaded container                              container=d05073c99bb2391a9d734d0fdabde12aed948f6bed20867e3b74e49c296f6ec9 paused=false running=true
DEBU[2021-01-16T10:41:01.955460046+01:00] loaded container                              container=9ce279b5b11c8bbdee1cbf3b5df0589482866ca65ecb7cf4d95200f4ae17f0cf paused=false running=true
DEBU[2021-01-16T10:41:01.955525296+01:00] loaded container                              container=b298a707a22475bb649577eca5118173fd6dfa4c256b433257a3ecc887299075 paused=false running=true
DEBU[2021-01-16T10:41:01.955586556+01:00] loaded container                              container=35b3511b5d031a7af1c9662b84e4b9cb419cc64f5be32ee029a62fd87142df4e paused=false running=true
DEBU[2021-01-16T10:41:01.955591966+01:00] loaded container                              container=040f4d98fd2fd53bfbf80c10f0e2cfa8430c3647c357925ac8c2c8ad39da5de5 paused=false running=false
DEBU[2021-01-16T10:41:01.955707287+01:00] loaded container                              container=b85216e818dcc4d0646eab7d10b59a4dd39764c74b893b1dc340893a916e7583 paused=false running=true
DEBU[2021-01-16T10:41:01.955724707+01:00] loaded container                              container=05925d61b119dadc2f92520fa7538800cb26759d0b201d5c032a386f26cb334f paused=false running=true
DEBU[2021-01-16T10:41:01.955829857+01:00] loaded container                              container=58e684e561d7f28a7bddf354b990ae2971741be6a8a70b1dc99d4c046cd16e46 paused=false running=true
DEBU[2021-01-16T10:41:01.955892417+01:00] loaded container                              container=ec35e488c3cc1d905d26f730d143c2e92a291d9c736e2997ce04d7cf2d93ca49 paused=false running=true
DEBU[2021-01-16T10:41:01.955903927+01:00] loaded container                              container=3b70cf99fbd840e2765e8fb21dc2329a6244a94b90132ad2c5a486827f3f0c40 paused=false running=false
DEBU[2021-01-16T10:41:01.955960287+01:00] loaded container                              container=5d4759a992cc4565ae57c72a04e29a46326bc4d1fef9818c9da9897cec417a3e paused=false running=true
DEBU[2021-01-16T10:41:01.956086257+01:00] loaded container                              container=46dee9c6e547dd01b266cb5d482a7ba38913e6a5a0f35c947f168974a4538f83 paused=false running=true
DEBU[2021-01-16T10:41:01.956098007+01:00] loaded container                              container=50a1b33282b79e1c3370b6ceb63d5424d028a999cabc7e56c5a810d14517d135 paused=false running=false
DEBU[2021-01-16T10:41:01.956218427+01:00] loaded container                              container=52c3067a761a0e4aa972cbb40ed5dda109ca1b4d3a78924be146b11a794689ed paused=false running=true
DEBU[2021-01-16T10:41:01.956337867+01:00] loaded container                              container=30d233415fdada278f09a2b02e2af64bcd960feda822eae6ef956eb5d8c0c8c9 paused=false running=true
DEBU[2021-01-16T10:41:01.956401747+01:00] loaded container                              container=9c5cb7c74ccd09c2de1816c23bc1bbc3356853520131bd74654888a907323b3d paused=false running=true
DEBU[2021-01-16T10:41:01.956422787+01:00] loaded container                              container=626a9fa32629d4d5e955e4c891e9292d3d6195ad670b6110b480c36560b5fa00 paused=false running=true
DEBU[2021-01-16T10:41:01.956436397+01:00] loaded container                              container=c83793206adcc0d62c685ea177ab62482ef1bd5bb7ee4ae112ba7ca2e1aa0937 paused=false running=true
DEBU[2021-01-16T10:41:01.956486107+01:00] loaded container                              container=1edfa3665f9edb5fb130c1d830ae81f2c6690e145a01c412254115d70a8ca95f paused=false running=true
DEBU[2021-01-16T10:41:01.956616027+01:00] loaded container                              container=7b0c0f00a9abf737db8fcf3c69ef18b2f5cb2614136621ebe5e2b4ea7849e59b paused=false running=true
DEBU[2021-01-16T10:41:01.956659807+01:00] loaded container                              container=f65c537189ccd6700a1c634b06e5fa3ab667a02599db2f9b83ceadd6de1f66db paused=false running=true
DEBU[2021-01-16T10:41:01.956695547+01:00] loaded container                              container=802dfac8ae9b0580f19d3f89d6a8dbb7057684abf11e9761cd5c496130a10b33 paused=false running=true
DEBU[2021-01-16T10:41:01.956871258+01:00] loaded container                              container=e8be58f56b32540a2eaa426344d4499b601cfc9877eaa8fbf86a7d81159eff05 paused=false running=true
DEBU[2021-01-16T10:41:01.978801850+01:00] restoring container                           container=b9bfe91a927adac0f9af1d0291374cf004b1045ec0c8d1186265aa579f2babab paused=false restarting=false running=true
DEBU[2021-01-16T10:41:01.985064477+01:00] restoring container                           container=9ce279b5b11c8bbdee1cbf3b5df0589482866ca65ecb7cf4d95200f4ae17f0cf paused=false restarting=false running=true
DEBU[2021-01-16T10:41:01.985242247+01:00] restoring container                           container=50a1b33282b79e1c3370b6ceb63d5424d028a999cabc7e56c5a810d14517d135 paused=false restarting=false running=false
DEBU[2021-01-16T10:41:01.985343087+01:00] restoring container                           container=ea9c118cc1d62286daccf23a4db5b7cc3a9eae78527a920ac33febffbfd924a6 paused=false restarting=false running=true
DEBU[2021-01-16T10:41:01.985559487+01:00] restoring container                           container=52c3067a761a0e4aa972cbb40ed5dda109ca1b4d3a78924be146b11a794689ed paused=false restarting=false running=true
DEBU[2021-01-16T10:41:01.985619407+01:00] restoring container                           container=3003e2bfed86be6550eda431ecae96cca300de0de5db6bb56864b65147f79dd6 paused=false restarting=false running=true
DEBU[2021-01-16T10:41:01.985998388+01:00] alive: false                                  container=50a1b33282b79e1c3370b6ceb63d5424d028a999cabc7e56c5a810d14517d135 paused=false restarting=false running=false
DEBU[2021-01-16T10:41:01.986026148+01:00] done restoring container                      container=50a1b33282b79e1c3370b6ceb63d5424d028a999cabc7e56c5a810d14517d135 paused=false restarting=false running=false
DEBU[2021-01-16T10:41:01.986581008+01:00] restoring container                           container=b85216e818dcc4d0646eab7d10b59a4dd39764c74b893b1dc340893a916e7583 paused=false restarting=false running=true
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x5586c892a7a4]

goroutine 420 [running]:
github.com/docker/docker/vendor/github.com/containerd/fifo.(*fifo).Close(0x0, 0x0, 0x0)
        /go/src/github.com/docker/docker/vendor/github.com/containerd/fifo/fifo.go:208 +0x44
github.com/docker/docker/vendor/github.com/containerd/containerd/cio.(*cio).Close(0xc000d06f60, 0x5586cb5654d0, 0xc000d8e9e8)
        /go/src/github.com/docker/docker/vendor/github.com/containerd/containerd/cio/io.go:203 +0x90
github.com/docker/docker/libcontainerd/remote.(*client).Restore.func1(0xc0008bf820, 0xc0008a2040)
        /go/src/github.com/docker/docker/libcontainerd/remote/client.go:86 +0x5a
github.com/docker/docker/libcontainerd/remote.(*client).Restore(0xc00098e5b0, 0x5586cb61c7c0, 0xc000052088, 0xc0011b6500, 0x40, 0xc0008bf810, 0x5586cb05cf00, 0xffffffffffffffff, 0x0, 0x0, ...)
        /go/src/github.com/docker/docker/libcontainerd/remote/client.go:107 +0x923
github.com/docker/docker/daemon.(*Daemon).restore.func3(0xc00079d9e0, 0xc000a38230, 0xc00000c1e0, 0xc00079d9a8, 0xc000d84f00, 0xc000d84ed0, 0xc000d84ea0, 0xc00128a280)
        /go/src/github.com/docker/docker/daemon/daemon.go:351 +0x48a
created by github.com/docker/docker/daemon.(*Daemon).restore
        /go/src/github.com/docker/docker/daemon/daemon.go:319 +0x4b3
```

### Steps to reproduce the behavior
Happend after the update to docker-ce:5:20.10.2~3-0~ubuntu-bionic


**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           20.10.2
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        2291f61
 Built:             Mon Dec 28 16:17:32 2020
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?

# dockerd --version
Docker version 20.10.2, build 8891c58
```

**Output of `docker info`:**

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.5.1-docker)

Server:
ERROR: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
errors pretty printing info
```

**Additional environment details (AWS, VirtualBox, physical, etc.)**

```
# uname -a
Linux server 5.10.7-051007-generic #202101122046 SMP Tue Jan 12 21:13:32 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

# containerd -v
containerd containerd.io 1.4.3 269548fa27e0089a8b8278fc4fc781d7f65a939b
```

--- a/vendor.mod
+++ b/vendor.mod
@@ -27,7 +27,7 @@ require (
 	github.com/containerd/cgroups v1.0.4
 	github.com/containerd/containerd v1.6.18
 	github.com/containerd/continuity v0.3.0
-	github.com/containerd/fifo v1.0.0
+	github.com/containerd/fifo v1.1.0
 	github.com/containerd/typeurl v1.0.2
 	github.com/coreos/go-systemd/v22 v22.4.0
 	github.com/creack/pty v1.1.11

--- a/vendor.sum
+++ b/vendor.sum
@@ -285,8 +285,9 @@ github.com/containerd/fifo v0.0.0-20190226154929-a9fb20d87448/go.mod h1:ODA38xgv
 github.com/containerd/fifo v0.0.0-20200410184934-f15a3290365b/go.mod h1:jPQ2IAeZRCYxpS/Cm1495vGFww6ecHmMk1YJH2Q5ln0=
 github.com/containerd/fifo v0.0.0-20201026212402-0724c46b320c/go.mod h1:jPQ2IAeZRCYxpS/Cm1495vGFww6ecHmMk1YJH2Q5ln0=
 github.com/containerd/fifo v0.0.0-20210316144830-115abcc95a1d/go.mod h1:ocF/ME1SX5b1AOlWi9r677YJmCPSwwWnQ9O123vzpE4=
-github.com/containerd/fifo v1.0.0 h1:6PirWBr9/L7GDamKr+XM0IeUFXu5mf3M/BPpH9gaLBU=
 github.com/containerd/fifo v1.0.0/go.mod h1:ocF/ME1SX5b1AOlWi9r677YJmCPSwwWnQ9O123vzpE4=
+github.com/containerd/fifo v1.1.0 h1:4I2mbh5stb1u6ycIABlBw9zgtlK8viPI9QkQNRQEEmY=
+github.com/containerd/fifo v1.1.0/go.mod h1:bmC4NWMbXlt2EZ0Hc7Fx7QzTFxgPID13eH0Qu+MAb2o=
 github.com/containerd/go-cni v1.0.1/go.mod h1:+vUpYxKvAF72G9i1WoDOiPGRtQpqsNW/ZHtSlv++smU=
 github.com/containerd/go-cni v1.0.2/go.mod h1:nrNABBHzu0ZwCug9Ije8hL2xBCYh/pjfMb1aZGrrohk=
 github.com/containerd/go-runc v0.0.0-20180907222934-5a6d9f37cfa3/go.mod h1:IV7qH3hrUgRmyYrtgEeGWJfWbgcHL9CSRruz2Vqcph0=

